sql/core/src/test/scala/org/apache/spark/sql/FileScanSuite.scala

```diff
@@ -0,0 +1,385 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import scala.collection.mutable
+
+import com.google.common.collect.ImmutableMap
+import org.apache.hadoop.fs.{FileStatus, Path}
+
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions.{And, Expression, IsNull, LessThan}
+import org.apache.spark.sql.execution.datasources.{PartitioningAwareFileIndex, PartitionSpec}
+import org.apache.spark.sql.execution.datasources.v2.FileScan
+import org.apache.spark.sql.execution.datasources.v2.csv.CSVScan
+import org.apache.spark.sql.execution.datasources.v2.json.JsonScan
+import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
+import org.apache.spark.sql.execution.datasources.v2.parquet.ParquetScan
+import org.apache.spark.sql.execution.datasources.v2.text.TextScan
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+trait FileScanSuiteBase extends SharedSparkSession {
+  private def newPartitioningAwareFileIndex() = {
+    new PartitioningAwareFileIndex(spark, Map.empty, None) {
+      override def partitionSpec(): PartitionSpec = {
+        PartitionSpec.emptySpec
+      }
+
+      override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = {
+        mutable.LinkedHashMap.empty
+      }
+
+      override protected def leafDirToChildrenFiles: Map[Path, Array[FileStatus]] = {
+        Map.empty
+      }
+
+      override def rootPaths: Seq[Path] = {
+        Seq.empty
+      }
+
+      override def refresh(): Unit = {}
+    }
+  }
+
+  type ScanBuilder = (
+    SparkSession,
+      PartitioningAwareFileIndex,
+      StructType,
+      StructType,
+      StructType,
+      Array[Filter],
+      CaseInsensitiveStringMap,
+      Seq[Expression],
+      Seq[Expression]) => FileScan
+
+  def run(scanBuilders: Seq[(String, ScanBuilder, Seq[String])]): Unit = {
+    val dataSchema = StructType(Seq(
+      StructField("data", IntegerType, false),
+      StructField("partition", IntegerType, false),
+      StructField("other", IntegerType, false)))
+    val dataSchemaNotEqual = StructType(Seq(
+      StructField("data", IntegerType, false),
+      StructField("partition", IntegerType, false),
+      StructField("other", IntegerType, false),
+      StructField("new", IntegerType, false)))
+    val readDataSchema = StructType(Seq(StructField("data", IntegerType, false)))
+    val readDataSchemaNotEqual = StructType(Seq(
+      StructField("data", IntegerType, false),
+      StructField("other", IntegerType, false)))
+    val readPartitionSchema = StructType(Seq(StructField("partition", IntegerType, false)))
+    val readPartitionSchemaNotEqual = StructType(Seq(
+      StructField("partition", IntegerType, false),
+      StructField("other", IntegerType, false)))
```

Comment:
In this case, shorter is better. Could you use the following style?
```scala
-    val dataSchema = StructType(Seq(
-      StructField("data", IntegerType, false),
-      StructField("partition", IntegerType, false),
-      StructField("other", IntegerType, false)))
-    val dataSchemaNotEqual = StructType(Seq(
-      StructField("data", IntegerType, false),
-      StructField("partition", IntegerType, false),
-      StructField("other", IntegerType, false),
-      StructField("new", IntegerType, false)))
-    val readDataSchema = StructType(Seq(StructField("data", IntegerType, false)))
-    val readDataSchemaNotEqual = StructType(Seq(
-      StructField("data", IntegerType, false),
-      StructField("other", IntegerType, false)))
-    val readPartitionSchema = StructType(Seq(StructField("partition", IntegerType, false)))
-    val readPartitionSchemaNotEqual = StructType(Seq(
-      StructField("partition", IntegerType, false),
-      StructField("other", IntegerType, false)))
+    val dataSchema = StructType.fromDDL("data INT, partition INT, other INT")
+    val dataSchemaNotEqual = StructType.fromDDL("data INT, partition INT, other INT, new INT")
+    val readDataSchema = StructType.fromDDL("data INT")
+    val readDataSchemaNotEqual = StructType.fromDDL("data INT, other INT")
+    val readPartitionSchema = StructType.fromDDL("partition INT")
+    val readPartitionSchemaNotEqual = StructType.fromDDL("partition INT, other INT")
```

Reply:
Thanks @dongjoon-hyun, I updated the PR with this change in: https://github.com/apache/spark/pull/31848/commits/d782723ee5728f54ad8884dab45e3d263a5fa442